libzfs_core/_error_translation.py

```diff
@@ -0,0 +1,469 @@
+"""
+
+import errno
+import os
+import re
+import string
+from . import exceptions as lzc_exc
+from ._constants import MAXNAMELEN
+
+
+def lzc_create_xlate_error(ret, name, is_zvol, props):
+    if ret == 0:
+        return
+    if ret == errno.EINVAL:
+        if not _is_valid_fs_name(name):
+            raise lzc_exc.NameInvalid(name)
+        elif len(name) > MAXNAMELEN:
+            raise lzc_exc.NameTooLong(name)
+        else:
+            raise lzc_exc.PropertyInvalid(name)
+
+    if ret == errno.EEXIST:
+        raise lzc_exc.FilesystemExists(name)
+    if ret == errno.ENOENT:
+        raise lzc_exc.ParentNotFound(name)
+    raise _generic_exception(ret, name, "Failed to create filesystem")
+
+
+def lzc_clone_xlate_error(ret, name, origin, props):
+    if ret == 0:
+        return
+    if ret == errno.EINVAL:
+        if not _is_valid_fs_name(name):
+            raise lzc_exc.NameInvalid(name)
+        elif not _is_valid_snap_name(origin):
+            raise lzc_exc.NameInvalid(origin)
+        elif len(name) > MAXNAMELEN:
+            raise lzc_exc.NameTooLong(name)
+        elif len(origin) > MAXNAMELEN:
+            raise lzc_exc.NameTooLong(origin)
+        elif _pool_name(name) != _pool_name(origin):
+            raise lzc_exc.PoolsDiffer(name) # see https://www.illumos.org/issues/5824
+        else:
+            raise lzc_exc.PropertyInvalid(name)
+
+    if ret == errno.EEXIST:
+        raise lzc_exc.FilesystemExists(name)
+    if ret == errno.ENOENT:
+        raise lzc_exc.DatasetNotFound(name)
+    raise _generic_exception(ret, name, "Failed to create clone")
+
+
+def lzc_rollback_xlate_error(ret, name):
+    if ret == 0:
+        return
+    if ret == errno.EINVAL:
+        if not _is_valid_fs_name(name):
+            raise lzc_exc.NameInvalid(name)
+        elif len(name) > MAXNAMELEN:
+            raise lzc_exc.NameTooLong(name)
+        else:
+            raise lzc_exc.SnapshotNotFound(name)
+    if ret == errno.ENOENT:
+        if not _is_valid_fs_name(name):
+            raise lzc_exc.NameInvalid(name)
+        else:
+            raise lzc_exc.FilesystemNotFound(name)
+    raise _generic_exception(ret, name, "Failed to rollback")
+
+
+def lzc_snapshot_xlate_errors(ret, errlist, snaps, props):
+    if ret == 0:
+        return
+
+    def _map(ret, name):
+        if ret == errno.EXDEV:
+            pool_names = map(_pool_name, snaps)
+            same_pool = all(x == pool_names[0] for x in pool_names)
+            if same_pool:
+                return lzc_exc.DuplicateSnapshots(name)
+            else:
+                return lzc_exc.PoolsDiffer(name)
+        elif ret == errno.EINVAL:
+            if any(not _is_valid_snap_name(s) for s in snaps):
+                return lzc_exc.NameInvalid(name)
+            elif any(len(s) > MAXNAMELEN for s in snaps):
+                return lzc_exc.NameTooLong(name)
+            else:
+                return lzc_exc.PropertyInvalid(name)
```

Comment:
> Perhaps, our next big ZFS project, if any, could be to introduce ZFS error codes that do not lose information (zfs_error_t) as opposed to the standard POSIX error codes.

That seems like it would be a good thing, yea.